<h1>Model Training</h1>

<h3>Import Data and Required Packages</h3>
<p>Import PANDAS, NUMPY, MATPLOTLIB, SEABORN, WARNINGS</p>



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

<h3>Import CSV Data as Pandas DataFrame</h3>

In [2]:
df = pd.read_csv('data/student_depression_dataset.csv')

***Showing Top 5 and Bottom 5 ROWS***

In [3]:
df.head()

id  Gender   Age           City Profession  Academic Pressure  \
0   2    Male  33.0  Visakhapatnam    Student                5.0   
1   8  Female  24.0      Bangalore    Student                2.0   
2  26    Male  31.0       Srinagar    Student                3.0   
3  30  Female  28.0       Varanasi    Student                3.0   
4  32  Female  25.0         Jaipur    Student                4.0   

   Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0            0.0  8.97                 2.0               0.0   
1            0.0  5.90                 5.0               0.0   
2            0.0  7.03                 5.0               0.0   
3            0.0  5.59                 2.0               0.0   
4            0.0  8.13                 3.0               0.0   

        Sleep Duration Dietary Habits   Degree  \
0          '5-6 hours'        Healthy  B.Pharm   
1          '5-6 hours'       Moderate      BSc   
2  'Less than 5 hours'        Healthy       BA   
3          '7-8 hours'       Moderate      BCA   
4          '5-6 hours'       Moderate   M.Tech   

  Have you ever had suicidal thoughts ?  Work/Study Hours Financial Stress  \
0                                   Yes               3.0              1.0   
1                                    No               3.0              2.0   
2                                    No               9.0              1.0   
3                                   Yes               4.0              5.0   
4                                   Yes               1.0              1.0   

  Family History of Mental Illness  Depression  
0                               No           1  
1                              Yes           0  
2                              Yes           0  
3                              Yes           1  
4                               No           0

In [4]:
df.tail()

id  Gender   Age       City Profession  Academic Pressure  \
27896  140685  Female  27.0      Surat    Student                5.0   
27897  140686    Male  27.0   Ludhiana    Student                2.0   
27898  140689    Male  31.0  Faridabad    Student                3.0   
27899  140690  Female  18.0   Ludhiana    Student                5.0   
27900  140699    Male  27.0      Patna    Student                4.0   

       Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
27896            0.0  5.75                 5.0               0.0   
27897            0.0  9.40                 3.0               0.0   
27898            0.0  6.61                 4.0               0.0   
27899            0.0  6.88                 2.0               0.0   
27900            0.0  9.24                 1.0               0.0   

            Sleep Duration Dietary Habits      Degree  \
27896          '5-6 hours'      Unhealthy  'Class 12'   
27897  'Less than 5 hours'        Healthy         MSc   
27898          '5-6 hours'      Unhealthy          MD   
27899  'Less than 5 hours'        Healthy  'Class 12'   
27900  'Less than 5 hours'        Healthy         BCA   

      Have you ever had suicidal thoughts ?  Work/Study Hours  \
27896                                   Yes               7.0   
27897                                    No               0.0   
27898                                    No              12.0   
27899                                   Yes              10.0   
27900                                   Yes               2.0   

      Financial Stress Family History of Mental Illness  Depression  
27896              1.0                              Yes           0  
27897              3.0                              Yes           0  
27898              2.0                               No           0  
27899              5.0                               No           1  
27900              3.0                              Yes           1

***Information About the Dataset***

In [5]:
df.describe()

id           Age  Academic Pressure  Work Pressure  \
count   27901.000000  27901.000000       27901.000000   27901.000000   
mean    70442.149421     25.822300           3.141214       0.000430   
std     40641.175216      4.905687           1.381465       0.043992   
min         2.000000     18.000000           0.000000       0.000000   
25%     35039.000000     21.000000           2.000000       0.000000   
50%     70684.000000     25.000000           3.000000       0.000000   
75%    105818.000000     30.000000           4.000000       0.000000   
max    140699.000000     59.000000           5.000000       5.000000   

               CGPA  Study Satisfaction  Job Satisfaction  Work/Study Hours  \
count  27901.000000        27901.000000      27901.000000      27901.000000   
mean       7.656104            2.943837          0.000681          7.156984   
std        1.470707            1.361148          0.044394          3.707642   
min        0.000000            0.000000          0.000000          0.000000   
25%        6.290000            2.000000          0.000000          4.000000   
50%        7.770000            3.000000          0.000000          8.000000   
75%        8.920000            4.000000          0.000000         10.000000   
max       10.000000            5.000000          4.000000         12.000000   

         Depression  
count  27901.000000  
mean       0.585499  
std        0.492645  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        1.000000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27901 entries, 0 to 27900
Data columns (total 18 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   id                                     27901 non-null  int64  
 1   Gender                                 27901 non-null  object 
 2   Age                                    27901 non-null  float64
 3   City                                   27901 non-null  object 
 4   Profession                             27901 non-null  object 
 5   Academic Pressure                      27901 non-null  float64
 6   Work Pressure                          27901 non-null  float64
 7   CGPA                                   27901 non-null  float64
 8   Study Satisfaction                     27901 non-null  float64
 9   Job Satisfaction                       27901 non-null  float64
 10  Sleep Duration                         27901 non-null  object 
 11  Di

In [7]:
df.shape

(27901, 18)

In [8]:
df.dtypes

id                                         int64
Gender                                    object
Age                                      float64
City                                      object
Profession                                object
Academic Pressure                        float64
Work Pressure                            float64
CGPA                                     float64
Study Satisfaction                       float64
Job Satisfaction                         float64
Sleep Duration                            object
Dietary Habits                            object
Degree                                    object
Have you ever had suicidal thoughts ?     object
Work/Study Hours                         float64
Financial Stress                          object
Family History of Mental Illness          object
Depression                                 int64
dtype: object

In [9]:
# Dropped the ID and CITY columns
df.drop(columns=['id', 'City'], inplace=True, axis=1)

In [10]:
columns = df.columns
columns

Index(['Gender', 'Age', 'Profession', 'Academic Pressure', 'Work Pressure',
       'CGPA', 'Study Satisfaction', 'Job Satisfaction', 'Sleep Duration',
       'Dietary Habits', 'Degree', 'Have you ever had suicidal thoughts ?',
       'Work/Study Hours', 'Financial Stress',
       'Family History of Mental Illness', 'Depression'],
      dtype='object')

In [11]:
df[columns].nunique()

Gender                                     2
Age                                       34
Profession                                14
Academic Pressure                          6
Work Pressure                              3
CGPA                                     332
Study Satisfaction                         6
Job Satisfaction                           5
Sleep Duration                             5
Dietary Habits                             4
Degree                                    28
Have you ever had suicidal thoughts ?      2
Work/Study Hours                          13
Financial Stress                           6
Family History of Mental Illness           2
Depression                                 2
dtype: int64

In [12]:
for column in columns:
    print(column)
    print(df[column].unique())

Gender
['Male' 'Female']
Age
[33. 24. 31. 28. 25. 29. 30. 27. 19. 20. 23. 18. 21. 22. 34. 32. 26. 39.
 35. 42. 36. 58. 49. 38. 51. 44. 43. 46. 59. 54. 48. 56. 37. 41.]
Profession
['Student' "'Civil Engineer'" 'Architect' "'UX/UI Designer'"
 "'Digital Marketer'" "'Content Writer'" "'Educational Consultant'"
 'Teacher' 'Manager' 'Chef' 'Doctor' 'Lawyer' 'Entrepreneur' 'Pharmacist']
Academic Pressure
[5. 2. 3. 4. 1. 0.]
Work Pressure
[0. 5. 2.]
CGPA
[ 8.97    5.9     7.03    5.59    8.13    5.7     9.54    8.04    9.79
  8.38    6.1     7.04    8.52    5.64    8.58    6.51    7.25    7.83
  9.93    8.74    6.73    5.57    8.59    7.1     6.08    5.74    9.86
  6.7     6.21    5.87    6.37    9.72    5.88    9.56    6.99    5.24
  9.21    7.85    6.95    5.86    7.92    9.66    8.94    9.71    7.87
  5.6     7.9     5.46    6.79    8.7     7.38    8.5     7.09    9.82
  8.89    7.94    9.11    6.75    7.53    9.49    9.01    7.64    5.27
  6.      9.44    5.75    7.51    9.05    6.38    8.

***Checking and NULL And Missing Values and Filling them***

In [13]:
df.isnull()

Gender    Age  Profession  Academic Pressure  Work Pressure   CGPA  \
0       False  False       False              False          False  False   
1       False  False       False              False          False  False   
2       False  False       False              False          False  False   
3       False  False       False              False          False  False   
4       False  False       False              False          False  False   
...       ...    ...         ...                ...            ...    ...   
27896   False  False       False              False          False  False   
27897   False  False       False              False          False  False   
27898   False  False       False              False          False  False   
27899   False  False       False              False          False  False   
27900   False  False       False              False          False  False   

       Study Satisfaction  Job Satisfaction  Sleep Duration  Dietary Habits  \
0                   False             False           False           False   
1                   False             False           False           False   
2                   False             False           False           False   
3                   False             False           False           False   
4                   False             False           False           False   
...                   ...               ...             ...             ...   
27896               False             False           False           False   
27897               False             False           False           False   
27898               False             False           False           False   
27899               False             False           False           False   
27900               False             False           False           False   

       Degree  Have you ever had suicidal thoughts ?  Work/Study Hours  \
0       False                                  False             False   
1       False                                  False             False   
2       False                                  False             False   
3       False                                  False             False   
4       False                                  False             False   
...       ...                                    ...               ...   
27896   False                                  False             False   
27897   False                                  False             False   
27898   False                                  False             False   
27899   False                                  False             False   
27900   False                                  False             False   

       Financial Stress  Family History of Mental Illness  Depression  
0                 False                             False       False  
1                 False                             False       False  
2                 False                             False       False  
3                 False                             False       False  
4                 False                             False       False  
...                 ...                               ...         ...  
27896             False                             False       False  
27897             False                             False       False  
27898             False                             False       False  
27899             False                             False       False  
27900             False                             False       False  

[27901 rows x 16 columns]

In [14]:
df.isnull().sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [15]:
df.isna().any()

Gender                                   False
Age                                      False
Profession                               False
Academic Pressure                        False
Work Pressure                            False
CGPA                                     False
Study Satisfaction                       False
Job Satisfaction                         False
Sleep Duration                           False
Dietary Habits                           False
Degree                                   False
Have you ever had suicidal thoughts ?    False
Work/Study Hours                         False
Financial Stress                         False
Family History of Mental Illness         False
Depression                               False
dtype: bool

In [16]:
(df == " ").sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [17]:
df.isin(['NA', 'NULL', '-', '?']).sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         3
Family History of Mental Illness         0
Depression                               0
dtype: int64

<p>I found 3 missing values with "?" in Financial Stress Column</p>

In [18]:
# Now Handling the Missing values

# Replacing the ? value with the NaN
df['Financial Stress'].replace('?', np.nan, inplace=True)

# Filling the NaN value with the MODE Imputer coz, its a categorical Column
df['Financial Stress'].fillna(df['Financial Stress'].mode()[0], inplace=True)

# Converting the Categorical column to Numerical Column
df['Financial Stress']=df['Financial Stress'].astype('float')


In [19]:
df.isin(['NA', 'NULL', '-', '?']).sum()

Gender                                   0
Age                                      0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [20]:
print(df['Financial Stress'].unique())
print(df['Financial Stress'].value_counts())

[1. 2. 5. 3. 4.]
Financial Stress
5.0    6718
4.0    5775
3.0    5226
1.0    5121
2.0    5061
Name: count, dtype: int64


***Retriving the Numerical Columns***

In [21]:
numerical_columns = df.select_dtypes(include=['number']).columns.tolist()
numerical_columns

['Age',
 'Academic Pressure',
 'Work Pressure',
 'CGPA',
 'Study Satisfaction',
 'Job Satisfaction',
 'Work/Study Hours',
 'Financial Stress',
 'Depression']

In [22]:
categorical_columns = df.select_dtypes(include=['object']).columns.tolist()
categorical_columns

['Gender',
 'Profession',
 'Sleep Duration',
 'Dietary Habits',
 'Degree',
 'Have you ever had suicidal thoughts ?',
 'Family History of Mental Illness']

In [23]:
for column in categorical_columns:
    print(column)
    print(df[column].unique())

Gender
['Male' 'Female']
Profession
['Student' "'Civil Engineer'" 'Architect' "'UX/UI Designer'"
 "'Digital Marketer'" "'Content Writer'" "'Educational Consultant'"
 'Teacher' 'Manager' 'Chef' 'Doctor' 'Lawyer' 'Entrepreneur' 'Pharmacist']
Sleep Duration
["'5-6 hours'" "'Less than 5 hours'" "'7-8 hours'" "'More than 8 hours'"
 'Others']
Dietary Habits
['Healthy' 'Moderate' 'Unhealthy' 'Others']
Degree
['B.Pharm' 'BSc' 'BA' 'BCA' 'M.Tech' 'PhD' "'Class 12'" 'B.Ed' 'LLB' 'BE'
 'M.Ed' 'MSc' 'BHM' 'M.Pharm' 'MCA' 'MA' 'B.Com' 'MD' 'MBA' 'MBBS' 'M.Com'
 'B.Arch' 'LLM' 'B.Tech' 'BBA' 'ME' 'MHM' 'Others']
Have you ever had suicidal thoughts ?
['Yes' 'No']
Family History of Mental Illness
['No' 'Yes']


In [24]:
df.shape

(27901, 16)

<h3>Preparing X varible and Y variable</h3>

<h3>1. Encoding</h3>
<table>
  <tr>
    <th>Column Name</th>
    <th>Encoding Method</th>
    <th>Reason</th>
  </tr>
  <tr>
    <td>Gender</td>
    <td>One-Hot Encoding ✅</td>
    <td>Nominal feature with limited categories (Male, Female)</td>
  </tr>
  <tr>
    <td>Profession</td>
    <td>Target Encoding ✅</td>
    <td>High cardinality; encoding based on target relationship</td>
  </tr>
  <tr>
    <td>Sleep Duration</td>
    <td>Ordinal Encoding ✅</td>
    <td>Ordered categories (Less than 5 hours &lt; 5-6 hours &lt; 7-8 hours &lt; ...)</td>
  </tr>
  <tr>
    <td>Dietary Habits</td>
    <td>Ordinal Encoding ✅</td>
    <td>Ordered categories (Healthy &gt; Moderate &gt; Unhealthy &gt; Others)</td>
  </tr>
  <tr>
    <td>Degree</td>
    <td>Target Encoding ✅</td>
    <td>High cardinality; encoding based on target relationship</td>
  </tr>
  <tr>
    <td>Have you ever had suicidal thoughts ?</td>
    <td>Ordinal Encoding ✅</td>
    <td>Binary values (Yes = 1, No = 0); encoded as ordered</td>
  </tr>
  <tr>
    <td>Family History of Mental Illness</td>
    <td>Ordinal Encoding ✅</td>
    <td>Binary values (Yes = 1, No = 0); encoded as ordered</td>
  </tr>
</table>


<h3>2. Scaling Numerical Features</h3>
<table>
  <tr>
    <th>Column Name</th>
    <th>Scaling Method</th>
    <th>Reason</th>
  </tr>
  <tr>
    <td>Age</td>
    <td>StandardScaler ✅</td>
    <td>Continuous numerical feature</td>
  </tr>
  <tr>
    <td>Academic Pressure</td>
    <td>StandardScaler ✅</td>
    <td>Numeric scale; normalization helps model performance</td>
  </tr>
  <tr>
    <td>Work Pressure</td>
    <td>StandardScaler ✅</td>
    <td>Continuous feature requiring normalization</td>
  </tr>
  <tr>
    <td>CGPA</td>
    <td>StandardScaler ✅</td>
    <td>Numerical range needs standardization</td>
  </tr>
  <tr>
    <td>Study Satisfaction</td>
    <td>StandardScaler ✅</td>
    <td>Numeric scale (e.g., Likert); better scaled</td>
  </tr>
  <tr>
    <td>Job Satisfaction</td>
    <td>StandardScaler ✅</td>
    <td>Similar to study satisfaction, likely ordinal scale</td>
  </tr>
  <tr>
    <td>Work/Study Hours</td>
    <td>StandardScaler ✅</td>
    <td>Continuous numerical feature</td>
  </tr>
  <tr>
    <td>Financial Stress</td>
    <td>StandardScaler ✅</td>
    <td>Numeric range; models benefit from scaling</td>
  </tr>
  <tr>
    <td>Degree (Target Encoded)</td>
    <td>StandardScaler ✅</td>
    <td>Target encoded numeric value; scaling needed</td>
  </tr>
  <tr>
    <td>Profession (Target Encoded)</td>
    <td>StandardScaler ✅</td>
    <td>Target encoded numeric value; scaling needed</td>
  </tr>
</table>


In [25]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
import category_encoders as ce


target_column = 'Depression'
cat_target_encoding_features = ['Degree', 'Profession']

# Apply Target Encoding
target_encoder = ce.TargetEncoder(handle_missing='value', handle_unknown='value')
df[cat_target_encoding_features] = target_encoder.fit_transform(
    df[cat_target_encoding_features], df[target_column]
)

# Define columns
numerical_columns = [
    'Age', 'Academic Pressure', 'Work Pressure', 'CGPA', 'Study Satisfaction',
    'Job Satisfaction', 'Work/Study Hours', 'Financial Stress', 'Degree', 'Profession'
]
cat_nominal_features = ['Gender']
cat_ordinal_features = ["Sleep Duration", "Dietary Habits"]
binary_features = ["Have you ever had suicidal thoughts ?", "Family History of Mental Illness"]

# Define categorical ordering
ordinal_categories = [
    ['Less than 5 hours', '5-6 hours', '7-8 hours', 'More than 8 hours', 'Others'],
    ['Healthy', 'Moderate', 'Unhealthy', 'Others']
]

# Pipelines
num_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

cat_nominal_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('one_hot_encoder', OneHotEncoder(handle_unknown='ignore'))
])

cat_ordinal_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal_encoder', OrdinalEncoder(categories=ordinal_categories, handle_unknown='use_encoded_value', unknown_value=-1)),
    ('scaler', StandardScaler())
])

cat_binary_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('binary_encoder', OrdinalEncoder())
])

# ColumnTransformer
preprocessor = ColumnTransformer(transformers=[
    ('num_pipeline', num_pipeline, numerical_columns),
    ('cat_nominal_pipeline', cat_nominal_pipeline, cat_nominal_features),
    ('cat_ordinal_pipeline', cat_ordinal_pipeline, cat_ordinal_features),
    ('cat_binary_pipeline', cat_binary_pipeline, binary_features)
], remainder='passthrough')

# Split
X = df.drop(columns=[target_column])
y = df[target_column]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Transform
X_train_transformed = preprocessor.fit_transform(X_train)
X_test_transformed = preprocessor.transform(X_test)


In [27]:
X_train.shape, X_test.shape

((22320, 15), (5581, 15))

In [28]:
X_train_transformed.shape, X_test_transformed.shape

((22320, 16), (5581, 16))

<h2>Machine Learning Models</h2>

In [42]:
## Train a Decision Tree Model 
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, recall_score, fbeta_score

# Initialize and Train a model
dt_model = DecisionTreeClassifier(random_state = 42)
dt_model.fit(X_train_transformed, y_train)

# Make Predictions
y_pred_dt = dt_model.predict(X_test_transformed)

# Evaluate the model
accuracy_dt = accuracy_score(y_test, y_pred_dt)
recall_dt = recall_score(y_test, y_pred_dt)
f2_score_dt = fbeta_score(y_test, y_pred_dt, beta=2)
print(f"Decision Tree Model Accuracy : {accuracy_dt*100:.2f}")
print(f"Decision Tree Model Recall : {recall_dt*100:.2f}")
print(f"Decision Tree Model f2_score : {f2_score_dt*100:.2f}")


Decision Tree Model Accuracy : 75.76
Decision Tree Model Recall : 78.57
Decision Tree Model f2_score : 78.74


In [43]:
## Train Logistic Regression Model
from sklearn.linear_model import LogisticRegression

# Initialize and Train a Model
lr_model = LogisticRegression()
lr_model.fit(X_train_transformed, y_train)

# Make Prediction
y_pred_lr = lr_model.predict(X_test_transformed)

# Evaluate the Model
accuracy_lr = accuracy_score(y_test, y_pred_lr)
recall_lr = recall_score(y_test, y_pred_lr)
f2_score_lr = fbeta_score(y_test, y_pred_lr, beta=2)
print(f"Logistic Regression Model Accuracy : {accuracy_lr*100:.2f}")
print(f"Logistic Regression Model Recall : {recall_lr*100:.2f}")
print(f"Logistic Regression Model f2_score : {f2_score_lr*100:.2f}")


Logistic Regression Model Accuracy : 83.84
Logistic Regression Model Recall : 87.52
Logistic Regression Model f2_score : 87.02


In [31]:
## Train Random Forest Model
from sklearn.ensemble import RandomForestClassifier

# Initialize and Train a model
rfc_model = RandomForestClassifier(n_estimators = 100, random_state = 42)
rfc_model.fit(X_train_transformed, y_train)

# Make Prediction
y_pred_rfc = rfc_model.predict(X_test_transformed)

# Evaluate the model
accuracy_rfc = accuracy_score(y_test, y_pred_rfc)
recall_rfc = recall_score(y_test, y_pred_rfc)
f2_score_rfc = fbeta_score(y_test, y_pred_rfc, beta=2)
print(f"Random Forest Classifier Model Accuracy : {accuracy_rfc*100:.2f}")
print(f"Random Forest Classifier Model Recall : {recall_rfc*100:.2f}")
print(f"Random Forest Classifier Model Recall : {f2_score_rfc*100:.2f}")


Random Forest Classifier Model Accuracy : 82.69
Random Forest Classifier Model Recall : 85.82
Random Forest Classifier Model Recall : 85.57


In [32]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from catboost import CatBoostClassifier


# Define Classifiers
model = {
    'SVM' : SVC(),
    'KNN' : KNeighborsClassifier(),
    'XGBoost' : XGBClassifier(),
    'CatBoost' : CatBoostClassifier(verbose = False),
    'AdaBoost' : AdaBoostClassifier(),
    'Gradient Boosting' : GradientBoostingClassifier()
}

# Train and Evaluate
for name, model in model.items():
    model.fit(X_train_transformed, y_train)
    y_pred = model.predict(X_test_transformed)
    acc = accuracy_score(y_test, y_pred)
    recall_all = recall_score(y_test, y_pred)
    f2_score_all = fbeta_score(y_test, y_pred, beta=2)
    print(f"{name} Accuracy : {acc*100:.2f}")
    print(f"{name} Recall Score -> {recall_all*100:.2f}")
    print(f"{name} f2 Score -> {f2_score_all*100:.2f}")

SVM Accuracy : 83.26
SVM Recall Score -> 87.86
SVM f2 Score -> 87.07
KNN Accuracy : 80.95
KNN Recall Score -> 86.38
KNN f2 Score -> 85.43
XGBoost Accuracy : 82.71
XGBoost Recall Score -> 86.10
XGBoost f2 Score -> 85.76
CatBoost Accuracy : 83.48
CatBoost Recall Score -> 87.15
CatBoost f2 Score -> 86.67
AdaBoost Accuracy : 84.02
AdaBoost Recall Score -> 87.68
AdaBoost f2 Score -> 87.17
Gradient Boosting Accuracy : 83.68
Gradient Boosting Recall Score -> 87.34
Gradient Boosting f2 Score -> 86.85


<h2>Model Performance Overview (Before Fine-Tuning)</h2>

<table border="1" cellpadding="5" cellspacing="0">
  <thead>
    <tr>
      <th>Model</th>
      <th>Accuracy (%)</th>
      <th>Recall (%)</th>
      <th>F2-score (%)</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>Decision Tree</td><td>75.76</td><td>78.57</td><td>78.74</td></tr>
    <tr><td>Logistic Regression</td><td>83.84</td><td>87.52</td><td>87.02</td></tr>
    <tr><td>Random Forest</td><td>82.69</td><td>85.82</td><td>85.57</td></tr>
    <tr><td>SVM</td><td>83.26</td><td>87.86</td><td>87.07</td></tr>
    <tr><td>KNN</td><td>80.95</td><td>86.38</td><td>85.43</td></tr>
    <tr><td>XGBoost</td><td>82.71</td><td>86.10</td><td>85.76</td></tr>
    <tr><td>CatBoost</td><td>83.48</td><td>87.15</td><td>86.67</td></tr>
    <tr><td>AdaBoost</td><td>84.02</td><td>87.68</td><td>87.17</td></tr>
    <tr><td>Gradient Boosting</td><td>83.68</td><td>87.34</td><td>86.85</td></tr>
  </tbody>
</table>

<h3>Why F2-score as Primary Metric?</h3>
<ul>
  <li><strong>F2-score</strong> emphasizes <strong>recall</strong> more than precision, which is vital in depression detection.</li>
  <li>In this context, missing a depressed student (<em>false negative</em>) is costlier than a false alarm.</li>
  <li>Optimizing for F2-score helps maximize true positive identification.</li>
</ul>

<h3>Insights</h3>
<ul>
  <li><strong>AdaBoost</strong> has the highest F2-score (87.17%) and strong accuracy (84.02%), making it a top candidate.</li>
  <li><strong>SVM</strong> and <strong>Logistic Regression</strong> follow closely with F2-scores above 87% and high recall.</li>
  <li><strong>Decision Tree</strong> performs the weakest overall and may need more tuning to improve.</li>
  <li>Other ensemble models (CatBoost, Gradient Boosting, XGBoost) also show competitive performance and robustness.</li>
</ul>

<h3>Next Steps: Fine-Tuning</h3>
<ul>
  <li>We will perform hyperparameter tuning on these models to improve F2-score and overall performance.</li>
  <li>Focus will be on top baseline models: AdaBoost, SVM, Logistic Regression, Gradient Boosting.</li>
  <li>Final model choice will balance recall-driven metrics and accuracy for deployment.</li>
</ul>


In [34]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, accuracy_score, recall_score, fbeta_score
import joblib

# Dictionary of classifiers (you can replace with your own model instances)
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

models = {
    "Logistic Regression": LogisticRegression(),
    "Random Forest": RandomForestClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "Support Vector Machine": SVC(),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB(),
    "AdaBoost": AdaBoostClassifier(),
    "Gradient Boosting": GradientBoostingClassifier()
}

# Your hyperparameter grid
params = {
    "Logistic Regression": {
        'C': [0.01, 0.1, 1, 10, 100],
        'penalty': ['l2'],
        'solver': ['liblinear', 'saga'],
        'class_weight': [None, 'balanced'],
        'max_iter': [1000]
    },
    "Random Forest": {
        'n_estimators': [100, 200],
        'max_depth': [None, 10, 20],
        'min_samples_split': [2, 5],
        'min_samples_leaf': [1, 2],
        'class_weight': [None, 'balanced']
    },
    "Decision Tree": {
        'criterion': ['gini', 'entropy'],
        'max_depth': [None, 10, 20],
        'min_samples_split': [2, 5],
        'min_samples_leaf': [1, 2],
        'class_weight': [None, 'balanced']
    },
    "Support Vector Machine": {
        'C': [0.1, 1, 10],
        'kernel': ['linear', 'rbf'],
        'gamma': ['scale', 'auto'],
        'class_weight': [None, 'balanced']
    },
    "K-Nearest Neighbors": {
        'n_neighbors': [3, 5, 7, 11],
        'weights': ['uniform', 'distance'],
        'metric': ['euclidean', 'manhattan']
    },
    "Naive Bayes": {},
    "AdaBoost": {
        'n_estimators': [100, 200],
        'learning_rate': [0.01, 0.1, 1]
    },
    "Gradient Boosting": {
        'n_estimators': [100, 200],
        'learning_rate': [0.01, 0.1],
        'subsample': [0.8, 1.0],
        'max_depth': [3, 5]
    }
}


# Define β for F2-score
BETA = 2

# Track best model based on F2-score
best_f2_score = 0
final_best_model = None
best_model_name = ""

best_models = {}

for name, model in models.items():
    print(f"\n\033[94mTuning {name}...\033[0m")
    param_grid = params[name]
    if param_grid:
        grid = GridSearchCV(model, param_grid, cv=5, scoring='recall', n_jobs=-1)
        grid.fit(X_train_transformed, y_train)
        best_model = grid.best_estimator_
        print(f"Best Params: {grid.best_params_}")
    else:
        model.fit(X_train_transformed, y_train)
        best_model = model
    
    best_models[name] = best_model
    y_pred = best_model.predict(X_test_transformed)

    acc = accuracy_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f2 = fbeta_score(y_test, y_pred, beta=BETA)

    print(classification_report(y_test, y_pred))
    print(f'{name} Model Accuracy Score : {acc*100:.5f}')
    print(f'{name} Model Recall Score : {recall*100:.5f}')
    print(f'{name} Model F2 Score : {f2*100:.5f}')

    # Update best model if F2 score is higher
    if f2 > best_f2_score:
        best_f2_score = f2
        final_best_model = best_model
        best_model_name = name

# ✅ Save best model and preprocessor
joblib.dump(final_best_model, "best_model.pkl")
joblib.dump(preprocessor, "preprocessor.pkl")

print(f"\n✅ Final Best Model Based on F2-score: {best_model_name}")
print("✅ Model and preprocessor saved as 'best_model.pkl' and 'preprocessor.pkl'")



Tuning Logistic Regression...
Best Params: {'C': 0.01, 'class_weight': None, 'max_iter': 1000, 'penalty': 'l2', 'solver': 'liblinear'}
              precision    recall  f1-score   support

           0       0.82      0.78      0.80      2343
           1       0.85      0.88      0.86      3238

    accuracy                           0.84      5581
   macro avg       0.84      0.83      0.83      5581
weighted avg       0.84      0.84      0.84      5581

Logistic Regression Model Accuracy Score : 83.76635
Logistic Regression Model Recall Score : 87.89376
Logistic Regression Model F2 Score : 87.23639

Tuning Random Forest...
Best Params: {'class_weight': None, 'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 200}
              precision    recall  f1-score   support

           0       0.82      0.78      0.80      2343
           1       0.84      0.87      0.86      3238

    accuracy                           0.83      5581
   macro avg       0.83  

<h2>Model Performance Report After Fine-Tuning</h2>

<p>After extensive hyperparameter tuning on various classification models for depression detection, we evaluated their performance using key metrics: Accuracy, Recall, and F2-score. The F2-score is prioritized here as it emphasizes recall, critical in mental health diagnosis to minimize false negatives.</p>

<table border="1" cellpadding="5" cellspacing="0" style="border-collapse: collapse; width: 100%;">
  <thead>
    <tr>
      <th>Model</th>
      <th>Best Parameters</th>
      <th>Accuracy (%)</th>
      <th>Recall (%)</th>
      <th>F2-score (%)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Logistic Regression</td>
      <td>C=0.01, penalty='l2', solver='liblinear', max_iter=1000</td>
      <td>83.77</td>
      <td>87.89</td>
      <td>87.24</td>
    </tr>
    <tr>
      <td>Random Forest</td>
      <td>max_depth=10, min_samples_leaf=2, n_estimators=200</td>
      <td>83.30</td>
      <td>87.46</td>
      <td>86.82</td>
    </tr>
    <tr>
      <td>Decision Tree</td>
      <td>max_depth=10, min_samples_split=5, criterion='gini'</td>
      <td>81.08</td>
      <td>84.65</td>
      <td>84.33</td>
    </tr>
    <tr>
      <td>Support Vector Machine (SVM)</td>
      <td>C=0.1, kernel='rbf', gamma='auto'</td>
      <td>83.80</td>
      <td>88.26</td>
      <td>87.49</td>
    </tr>
    <tr>
      <td>K-Nearest Neighbors (KNN)</td>
      <td>n_neighbors=11, metric='manhattan', weights='uniform'</td>
      <td>81.97</td>
      <td>87.92</td>
      <td>86.72</td>
    </tr>
    <tr>
      <td>Naive Bayes</td>
      <td>–</td>
      <td>82.55</td>
      <td>82.89</td>
      <td>83.58</td>
    </tr>
    <tr>
      <td>AdaBoost</td>
      <td>learning_rate=0.1, n_estimators=100</td>
      <td>82.92</td>
      <td>89.22</td>
      <td>87.84</td>
    </tr>
    <tr>
      <td>Gradient Boosting</td>
      <td>learning_rate=0.01, max_depth=5, n_estimators=100, subsample=0.8</td>
      <td>82.60</td>
      <td>89.56</td>
      <td><strong>87.96</strong></td>
    </tr>
  </tbody>
</table>

<h3>Key Insights</h3>
<ul>
  <li><strong>Gradient Boosting</strong> achieved the highest F2-score (87.96%), indicating the best balance favoring recall, crucial for depression detection.</li>
  <li><strong>AdaBoost</strong> and <strong>SVM</strong> also performed strongly, with F2-scores close to 87.8% and 87.5%, respectively.</li>
  <li><strong>Logistic Regression</strong> remains a strong baseline model with an F2-score of 87.24% and good interpretability.</li>
  <li><strong>Decision Tree</strong> and <strong>Naive Bayes</strong> had lower performance metrics, suggesting they may require further tuning or may not be optimal choices.</li>
  <li>Recall scores above 84% for all models demonstrate good sensitivity to depressed cases, reducing missed diagnoses.</li>
</ul>

<h3>Conclusion and Next Steps</h3>
<p>The <strong>Gradient Boosting model</strong> is selected as the <em>final best model</em> based on the F2-score, emphasizing high recall with balanced precision. Both the model and the preprocessor were saved for deployment as <code>best_model.pkl</code> and <code>preprocessor.pkl</code>.</p>

<p>Future work includes testing on external datasets, further hyperparameter optimization, and integrating this model into a user-friendly application to assist in early depression detection.</p>


In [35]:
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, recall_score

best_models = {}

for name, model in models.items():
    print(f"\n\033[94mTuning {name}...\033[0m")
    param_grid = params[name]
    
    if param_grid:
        grid = GridSearchCV(model, param_grid, cv=5, scoring='recall', n_jobs=-1)
        grid.fit(X_train_transformed, y_train)
        best_model = grid.best_estimator_
        print(f"Best Params: {grid.best_params_}")
    else:
        model.fit(X_train_transformed, y_train)
        best_model = model
    
    best_models[name] = best_model
    
    y_pred = best_model.predict(X_test_transformed)

    # Evaluation metrics
    acc = accuracy_score(y_test, y_pred)
    rec = recall_score(y_test, y_pred)
    f2 = fbeta_score(y_test, y_pred, beta=BETA)
    cm = confusion_matrix(y_test, y_pred)

    # Extract TP and FN
    tn, fp, fn, tp = cm.ravel()

    print("\nClassification Report:")
    print(classification_report(y_test, y_pred))
    
    print(f"{name} Model Accuracy Score : {acc*100:.5f}")
    print(f"{name} Model Recall Score   : {rec*100:.5f}")
    print(f"{name} Model f2 Score   : {f2*100:.5f}")
    print("Confusion Matrix:")
    print(cm)
    print(f"True Positives (TP): {tp}")
    print(f"False Negatives (FN): {fn}")



Tuning Logistic Regression...
Best Params: {'C': 0.01, 'class_weight': None, 'max_iter': 1000, 'penalty': 'l2', 'solver': 'liblinear'}

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.78      0.80      2343
           1       0.85      0.88      0.86      3238

    accuracy                           0.84      5581
   macro avg       0.84      0.83      0.83      5581
weighted avg       0.84      0.84      0.84      5581

Logistic Regression Model Accuracy Score : 83.76635
Logistic Regression Model Recall Score   : 87.89376
Logistic Regression Model f2 Score   : 87.23639
Confusion Matrix:
[[1829  514]
 [ 392 2846]]
True Positives (TP): 2846
False Negatives (FN): 392

Tuning Random Forest...
Best Params: {'class_weight': None, 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.78      

In [36]:
import joblib

joblib.dump(target_encoder, "target_encoder.pkl")
joblib.dump(preprocessor, "preprocessor.pkl")
joblib.dump(final_best_model, "best_model.pkl")


['best_model.pkl']

In [39]:
import pandas as pd
import joblib

# Load model and preprocessing components
model = joblib.load("best_model.pkl")
preprocessor = joblib.load("preprocessor.pkl")
target_encoder = joblib.load("target_encoder.pkl")

# Simulated user input
user_input = {
    "Gender": "Male",
    "Age": 27.0,
    "Profession": "Student",
    "Academic Pressure": 2.0,
    "Work Pressure": 0.0,
    "CGPA": 9.40,
    "Study Satisfaction": 3.0,
    "Job Satisfaction": 0.0,
    "Sleep Duration": "'Less than 5 hours'",  
    "Dietary Habits": "Healthy",
    "Degree": "MSc",
    "Have you ever had suicidal thoughts ?": "No",
    "Work/Study Hours": 0.0,
    "Financial Stress": 3.0,
    "Family History of Mental Illness": "Yes"
}

# Convert input to DataFrame
input_df = pd.DataFrame([user_input])

# Step 1: Target Encoding
target_encoding_features = ['Degree', 'Profession']
input_df[target_encoding_features] = target_encoder.transform(input_df[target_encoding_features])

# Step 2: ColumnTransformer (includes all other encoding + scaling)
transformed_input = preprocessor.transform(input_df)

# Step 3: Prediction
prediction = model.predict(transformed_input)[0]
probability = model.predict_proba(transformed_input)[0][1]

# Output
print("📊 Prediction:", "😞 Depressed" if prediction == 1 else "🙂 Not Depressed")
print(f"🧠 Probability of Depression: {probability * 100:.2f}%")


📊 Prediction: 🙂 Not Depressed
🧠 Probability of Depression: 27.11%


In [ ]:

import pandas as pd
import joblib

# Load model and preprocessor
model = joblib.load("best_model.pkl")
preprocessor = joblib.load("preprocessor.pkl")
#label_encoders = joblib.load("label_encoders.pkl")  # You must save this during training
target_encoder = joblib.load("target_encoder.pkl")  # You must save this during training

# Take input from user
user_input = {
    "Gender": input("Enter Gender (Male/Female): "),
    "Age": float(input("Enter Age: ")),
    "Profession": input("Enter Profession (e.g., Student): "),
    "Academic Pressure": float(input("Enter Academic Pressure (0-5): ")),
    "Work Pressure": float(input("Enter Work Pressure (0-5): ")),
    "CGPA": float(input("Enter CGPA: ")),
    "Study Satisfaction": float(input("Enter Study Satisfaction (0-6): ")),
    "Job Satisfaction": float(input("Enter Job Satisfaction (0-4): ")),
    "Sleep Duration": input("Enter Sleep Duration (e.g., 'Less than 5 hours'): "),  # Keep single quotes
    "Dietary Habits": input("Enter Dietary Habits (e.g., Healthy/Unhealthy): "),
    "Degree": input("Enter Degree (e.g., MSc, B.Tech): "),
    "Have you ever had suicidal thoughts ?": input("Have you ever had suicidal thoughts ? (Yes/No): "),
    "Work/Study Hours": float(input("Enter Work/Study Hours (0-12): ")),
    "Financial Stress": float(input("Enter Financial Stress (1-5): ")),
    "Family History of Mental Illness": input("Family History of Mental Illness? (Yes/No): ")
}

# Convert input to DataFrame
input_df = pd.DataFrame([user_input])

# Step 1: Target Encoding
target_encoding_features = ['Degree', 'Profession']
input_df[target_encoding_features] = target_encoder.transform(input_df[target_encoding_features])

# Step 2: ColumnTransformer (includes all other encoding + scaling)
transformed_input = preprocessor.transform(input_df)

# Step 4: Make Prediction
prediction = model.predict(transformed_input)[0]
probability = model.predict_proba(transformed_input)[0][1]

# Final Output with emojis
print("Prediction:", "😞 Depressed" if prediction == 1 else "🙂 Not Depressed")
print(f"Probability of Depression: {probability * 100:.2f}%")




Prediction: 😞 Depressed
Probability of Depression: 79.43%
